### PR TITLE
Fix panic in QueryDatastore when session becomes nil

### PR DESCRIPTION
**Pull request by gigawhitlocks**
When sessions are expired (or nuked via govc), we were seeing an inconsistent state in vicadmin caused by a panic in QueryDatastore. This fixes that by terminating QueryDatastore if the datastore is inaccessible. The method also returns an err now, though it's not necessary that we do anything with the error at call time currently.

### CODE REVIEW

File: lib/vicadmin/validate.go [hunk 1]
```diff
@@ -130,41 +130,41 @@ func NewValidator(ctx context.Context, vch *config.VirtualContainerHostConfigSpe
 		v.FirewallIssues = template.HTML("")
 		v.LicenseStatus = BadStatus
 		v.LicenseIssues = template.HTML("")
-	} else {
+		return nil
```
Comment (gigawhitlocks) [inline lib/vicadmin/validate.go]: This hunk looks like a bunch of changes but really all I did was add `return nil` to the end of the `if` statement and outdent the `else` block & remove the `else` keyword.

Comment (emlin) [inline lib/vicadmin/validate.go]: what about change to
```
if sess != nil {
   return nil
}
```

Comment (anchal-agrawal) [inline lib/vicadmin/validate.go]: Returning `nil` here causes a panic on https://github.com/vmware/vic/blob/master/cmd/vicadmin/server.go#L560 when the session is nil.

File: lib/vicadmin/validate.go [hunk 2]
```diff
@@ -258,7 +259,7 @@ func (v *Validator) QueryDatastore(ctx context.Context, vch *config.VirtualConta
 	sort.Sort(dataStores)
 	if err != nil {
 		log.Errorf("Error while accessing datastore: %s", err)
-		return
+		return err
```
Comment (emlin) [inline lib/vicadmin/validate.go]: curious for what are you gonna do with this returned error?

**Comment by emlin** [comment]
I think your return can only fix the panic caused by not authenticated issue, but not actually fix the problem. And potentially will miss other errors if that's not not authenticated issue happens during datastore query.

I have one suggestion to fix the issue. Cause you verified if the session creation time is longer than 24h while user session is retrieved from session cache. 
I think you can do a simply vsphere call using that session like what keepalive did in the session.go file. https://github.com/vmware/vic/blob/master/pkg/vsphere/session/session.go#L189
If you got not authenticated error, you might just return error similar to the session is already longer than 24h.
Then suppose the user will be forward to the login page.

File: lib/vicadmin/validate.go (hunk 2 above)
Comment (gigawhitlocks) [inline lib/vicadmin/validate.go]: Nothing, it's thrown out. It's just used to exit the function before we start trying to access `nil` members.

File: lib/vicadmin/validate.go (hunk 1 above)
Comment (gigawhitlocks) [inline lib/vicadmin/validate.go]: I'll just undo this refactor. It's not relevant to this PR.

File: lib/vicadmin/validate.go [hunk 3]
```diff
@@ -247,6 +247,7 @@ func (v *Validator) QueryDatastore(ctx context.Context, vch *config.VirtualConta
 		ds, err := sess.Finder.DatastoreOrDefault(ctx, dsName)
 		if err != nil {
 			log.Errorf("Unable to collect information for datastore %s: %s", dsName, err)
+			return err
```
Comment (gigawhitlocks) [inline lib/vicadmin/validate.go]: true

File: cmd/vicadmin/usersession.go [hunk 4]
```diff
@@ -89,6 +89,15 @@ func (u *UserSessionStore) VSphere(ctx context.Context, id string) (*session.Ses
 		return nil, fmt.Errorf("No vSphere session found. User with ID %s must authenticate again.", id)
 	}
 
+	vsphus, err := us.vsphere.SessionManager.UserSession(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("vSphere UserSession was invalid; got error: %s", err.Error())
+	}
+
+	if vsphus == nil {
+		return nil, fmt.Errorf("vSphere UserSession came back nil but no error was reported")
```
Comment (hmahmood) [inline cmd/vicadmin/usersession.go]: Is this error user visible?

Comment (gigawhitlocks) [inline cmd/vicadmin/usersession.go]: Shouldn't be but I'll make sure. I've been surprised in a few places when an error has bubbled up to the UI :\

**Comment by mdubya66** [comment]
what issue is this fixing?

File: cmd/vicadmin/vicadm.go [hunk 5]
```diff
@@ -483,6 +483,9 @@ func vSphereSessionGet(sessconfig *session.Config) (*session.Session, error) {
 		log.Errorf("Got %s while creating user session", err)
 		return nil, err
 	}
+	if usersession == nil {
+		return nil, fmt.Errorf("UserSession from vSphere came back nil but no error was reported")
```
Comment (dougm) [inline cmd/vicadmin/vicadm.go]: Not sure why you chose this error message.. as we discussed and as is documented, UserSession returns nil, nil if the session is no longer valid.  At which point you need to `Login()` to get a new session.

https://github.com/vmware/govmomi/blob/master/session/manager.go#L133-L134

File: lib/vicadmin/validate.go [hunk 6]
```diff
@@ -252,13 +255,25 @@ func (v *Validator) QueryDatastore(ctx context.Context, vch *config.VirtualConta
...
 
+	if sess.Client == nil {
+		return fmt.Errorf("Non-nil session govmomi client is nil while looking for datastore info")
+	}
+
+	if sess.Client.Client == nil {
+		return fmt.Errorf("Non-nil session vim client is nil while looking for datastore info")
+	}
+
 	pc := property.DefaultCollector(sess.Client.Client)
```
Comment (dougm) [inline lib/vicadmin/validate.go]: property.DefaultCollector never returns nil.

File: lib/vicadmin/validate.go [hunk 7]
```diff
@@ -252,13 +255,25 @@ func (v *Validator) QueryDatastore(ctx context.Context, vch *config.VirtualConta
 		}
 	}
 
+	if sess.Client == nil {
```
Comment (dougm) [inline lib/vicadmin/validate.go]: Where would this get set to nil?

File: lib/vicadmin/validate.go [hunk 8]
```diff
@@ -252,13 +255,25 @@ func (v *Validator) QueryDatastore(ctx context.Context, vch *config.VirtualConta
...
+		return fmt.Errorf("Non-nil session govmomi client is nil while looking for datastore info")
+	}
+
+	if sess.Client.Client == nil {
```
Comment (dougm) [inline lib/vicadmin/validate.go]: Where would this get set to nil?

File: lib/vicadmin/validate.go [hunk 9]
```diff
@@ -252,13 +255,25 @@ func (v *Validator) QueryDatastore(ctx context.Context, vch *config.VirtualConta
...
+		return fmt.Errorf("Non-nil session vim client is nil while looking for datastore info")
+	}
+
 	pc := property.DefaultCollector(sess.Client.Client)
+	if pc == nil {
+		return fmt.Errorf("Could not get default propery collector; prop-collector came back nil")
+	}
+
 	err := pc.Retrieve(ctx, refs, nil, &dataStores)
```
Comment (dougm) [inline lib/vicadmin/validate.go]: As we discussed in slack, the problem happens above, when looking up datastores with an invalid session.  The error is logged above
`log.Errorf("Unable to collect information for datastore %s: %s", dsName, err)`

But continues, and `refs` passed to `pc.Retrieve` has len == 0, causing the panic.  So the log.Errorf above should either be followed by a return error or you should check for `len(refs) == 0` here and avoid calling pc.Retrieve in that case.

File: lib/vicadmin/validate.go [hunk 10]
```diff
@@ -252,13 +255,25 @@ func (v *Validator) QueryDatastore(ctx context.Context, vch *config.VirtualConta
...
+	if sess.Client == nil {
+		return fmt.Errorf("Non-nil session govmomi client is nil while looking for datastore info")
+	}
+
+	if sess.Client.Client == nil {
+		return fmt.Errorf("Non-nil session vim client is nil while looking for datastore info")
+	}
+
 	pc := property.DefaultCollector(sess.Client.Client)
+	if pc == nil {
```
Comment (dougm) [inline lib/vicadmin/validate.go]: property.DefaultCollector never returns nil.

**Comment by gigawhitlocks** [comment]
@mdubya66 trying to fix the panics seen over the weekend -- idk if an issue was reported

File: lib/vicadmin/validate.go (hunk 9 above)
Comment (gigawhitlocks) [inline lib/vicadmin/validate.go]: Well I had a `return err` there originally but @emlin told me not to return err there, because there may be other datastores that are working. I misremembered and thought that the properycollector could come back nil, not `refs`

Comment (gigawhitlocks) [inline lib/vicadmin/validate.go]: I'll change the above nil checks to check refs for len 0 instead & return error otherwise.

**Comment by mdubya66** [comment]
#3258 and #3259 @gigawhitlocks

**Comment by dougm** [comment]
We haven't been using persistent connections in vicadmin.  If you switch to using a keepalive function, you'll be making the connection persistent too.  hence the suggestion to use the UserSession helper to determine if a session cookie is still valid (similar to how govc does).  When using keepalive func, there's no need to use the UserSession helper (port layer case for example)

Sent from my iPhone

On Nov 21, 2016, at 3:40 PM, Ian Whitlock <notifications@github.com<mailto:notifications@github.com>> wrote:


@gigawhitlocks commented on this pull request.

________________________________

In cmd/vicadmin/vicadm.go<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_vmware_vic_pull_3270&d=DgMCaQ&c=uilaK90D4TOVoH58JNXRgQ&r=GroxMYS-chy4r563y5jUmA&m=7EoSKTD5zMZ6n4JlE7U6v12FbEruMoDs9rxH-BUKk9g&s=Sj_r8arsZRsuphKknZs-EL_-xMr5bMgnVq2xMhuEmnI&e=>:

> @@ -483,6 +483,9 @@ func vSphereSessionGet(sessconfig *session.Config) (*session.Session, error) {
                log.Errorf("Got %s while creating user session", err)
                return nil, err
        }
+       if usersession == nil {
+               return nil, fmt.Errorf("UserSession from vSphere came back nil but no error was reported")


My understanding is that we should not need to call Login but I might need to register my own KeepAlive as the one that gets registered when I call Connect() doesn't look like it's actually getting registered because the KeepAlive value == 0, so I suspect that's what's causing the connection to go stale. And then we don't have a way, nor do we want a way, to establish a new session if the session goes stale.

-

**Comment by gigawhitlocks** [comment]
Got it, thought we were using persistent connections, thanks for clarifying @dougm .
I added a check to `usersession.VSphere` to try to use the `UserSession` helper to check to make sure that is valid and I'm testing that right now.

**Comment by gigawhitlocks** [comment]
Starting to get real tired tho :(

**Comment by gigawhitlocks** [comment]
OK I have the functionality working the way it should, but I'm doing it by comparing error messages, which is surely slow (though only happens when there is an error at all) and brittle, so I'm going to give it a little thought and come up with something better, but the PR does resolve the issue (tested by `vic-machine create` then log into vicadmin in the browser then use `govc` OOB to delete the sessions, then load the tab again (not Refresh, which will `POST` again and actually log in :P ) and we get directed to the login page. `Redirect` sends the browser to `logout` so that we can do cleanup.

**Comment by gigawhitlocks** [comment]
@mdubya66 #3259 is addressed by a single line in this change and the rest is for #3258

File: cmd/vicadmin/server.go [hunk 11]
```diff
@@ -439,13 +439,14 @@ func (s *server) bundleContainerLogs(res http.ResponseWriter, req *http.Request,
 	c, err := s.getSessionFromRequest(context.Background(), req)
 	if err != nil {
 		log.Errorf("Failed to get vSphere session while bundling container logs due to error: %s", err.Error())
+		if strings.Contains(err.Error(), "Could not validate session") {
```
Comment (vburenin) [inline cmd/vicadmin/server.go]: is there any chance message can be localized?

File: cmd/vicadmin/usersession.go [hunk 12]
```diff
@@ -86,9 +86,18 @@ func (u *UserSessionStore) VSphere(ctx context.Context, id string) (*session.Ses
 		return nil, fmt.Errorf("User session with unique ID %s does not exist", id)
 	}
 	if us.vsphere == nil {
-		return nil, fmt.Errorf("No vSphere session found. User with ID %s must authenticate again.", id)
+		return nil, fmt.Errorf("No vSphere session found for user with ID %s", id)
```
Comment (vburenin) [inline cmd/vicadmin/usersession.go]: for user: %s

File: cmd/vicadmin/usersession.go [hunk 13]
```diff
@@ -86,9 +86,18 @@ func (u *UserSessionStore) VSphere(ctx context.Context, id string) (*session.Ses
...
 	}
 
+	vsphus, err := us.vsphere.SessionManager.UserSession(ctx)
+	if err != nil || vsphus == nil {
+		u.Delete(id)
+		errMsg := fmt.Sprintf("Could not validate session for user %s; must log in again", id)
```
Comment (vburenin) [inline cmd/vicadmin/usersession.go]: Why do you even try to format this message not checking err first?

```
if err != nil {
    return nil, fmt.Errorf("%s. Got error %s", errMsg, err.Error())
}
return nil,  fmt.Errorf("Could not validate session for user %s; must log in again", id)
```

Comment (gigawhitlocks) [inline cmd/vicadmin/usersession.go]: I format the message w/o the error in the case where `err` == `nil` but `vsphus` is `nil`. If err is non-nil the err is appended. That is why I format first.

File: lib/vicadmin/validate.go [hunk 14]
```diff
@@ -253,12 +256,19 @@ func (v *Validator) QueryDatastore(ctx context.Context, vch *config.VirtualConta
 	}
 
 	pc := property.DefaultCollector(sess.Client.Client)
+	if pc == nil {
+		return fmt.Errorf("Could not get default propery collector; prop-collector came back nil")
+	}
+
+	if len(refs) == 0 {
```
Comment (cgtexmex) [inline lib/vicadmin/validate.go]: To improve readability I'd move this right after the call for references

File: lib/vicadmin/validate.go [hunk 15]
```diff
@@ -253,12 +256,19 @@ func (v *Validator) QueryDatastore(ctx context.Context, vch *config.VirtualConta
...
+		return fmt.Errorf("No datastore references found")
+	}
+
 	err := pc.Retrieve(ctx, refs, nil, &dataStores)
 
 	sort.Sort(dataStores)
 	if err != nil {
```
Comment (cgtexmex) [inline lib/vicadmin/validate.go]: Same as previous comment -- I'd move this up beneath the call to pc.Retrieve...

File: lib/vicadmin/validate.go [hunk 16]
```diff
@@ -209,7 +209,10 @@ func NewValidator(ctx context.Context, vch *config.VirtualContainerHostConfigSpe
 		v.DockerPort = fmt.Sprintf("%d", opts.DefaultTLSHTTPPort)
 	}
 
-	v.QueryDatastore(ctx, vch, sess)
+	err = v.QueryDatastore(ctx, vch, sess)
+	if err != nil {
+		log.Errorf("Had a problem querying the datastores: %s", err.Error())
```
Comment (gigawhitlocks) [inline lib/vicadmin/validate.go]: Yeah we just want to report because we want to continue to QueryVCHStatus and we still want to return `v` down at the bottom even if this step fails.

File: lib/vicadmin/validate.go [hunk 17]
```diff
@@ -253,12 +256,19 @@ func (v *Validator) QueryDatastore(ctx context.Context, vch *config.VirtualConta
...
+	}
+
 	err := pc.Retrieve(ctx, refs, nil, &dataStores)
 
 	sort.Sort(dataStores)
 	if err != nil {
 		log.Errorf("Error while accessing datastore: %s", err)
-		return
 	}
 	for _, ds := range dataStores {
```
Comment (gigawhitlocks) [inline lib/vicadmin/validate.go]: Yeah, we show that it was inaccessible in the UI

File: cmd/vicadmin/usersession.go (hunk 13 above)
Comment (vburenin) [inline cmd/vicadmin/usersession.go]: My point is that overall code organization could be better:
okay, let's reorganize it better:
```
if err != nil {
    return nil, fmt.Errorf("Failed to validate user %s session: %v", id, err)
}
return nil, fmt.Errorf("User %s session has expired", id)
```

File: lib/vicadmin/validate.go [hunk 18]
```diff
@@ -252,13 +255,19 @@ func (v *Validator) QueryDatastore(ctx context.Context, vch *config.VirtualConta
...
+		return fmt.Errorf("No datastore references found")
+	}
+
 	pc := property.DefaultCollector(sess.Client.Client)
-	err := pc.Retrieve(ctx, refs, nil, &dataStores)
+	if pc == nil {
+		return fmt.Errorf("Could not get default propery collector; prop-collector came back nil")
+	}
 
 	sort.Sort(dataStores)
```
Comment (cgtexmex) [inline lib/vicadmin/validate.go]: this needs to be after the pc.retrieve or the sort won't work...

**Comment by emlin** [comment]
The session valid checking is in getSessionFromRequest, and error is handled in index and bundleContainerLogs. So if the vsphere session is timeout, and if user access the <ip:2378>/ or the bundle log ip address, they will be redirected to login page.
But if user type portlayer log, persona log address directly, they don't need to relogin. That page will be available for 24hours.

You might test it with removed vsphere active session.

I think we're using vsphere account to login, and manage user session, we should make all URLs access same for session management. So idea is still to change the isAuthenticated method, to make sure access for any URLs rely on this session check.
@dougm @hickeng 

Another benefit is that will centralize all session related management and login/logout redirect logic. (my 2 cents)

**Comment by gigawhitlocks** [comment]
> But if user type portlayer log, persona log address directly, they don't need to relogin. That page will be available for 24hours.

This was the desired behavior requested by @hickeng and is not related to this PR as it is not changed by this PR

>I think we're using vsphere account to login, and manage user session, we should make all URLs access same for session management. So idea is still to change the isAuthenticated method, to make sure access for any URLs rely on this session check.

This was the behavior I implemented and then  @anchal-agrawal was told to change it to the current behavior.

File: lib/vicadmin/validate.go (hunk 18 above)
Comment (gigawhitlocks) [inline lib/vicadmin/validate.go]: oh, yeah, you're right, moving stuff around I definitely confused myself.

File: cmd/vicadmin/usersession.go (hunk 13 above)
Comment (gigawhitlocks) [inline cmd/vicadmin/usersession.go]: oh I see

**Comment by cgtexmex** [comment]
lgtm

File: cmd/vicadmin/server.go [hunk 19]
```diff
@@ -551,6 +548,10 @@ func (s *server) index(res http.ResponseWriter, req *http.Request) {
 	defer trace.End(trace.Begin(""))
 	ctx := context.Background()
 	sess, err := s.getSessionFromRequest(ctx, req)
+	if err != nil {
+		http.Redirect(res, req, "/logout", http.StatusTemporaryRedirect)
+		return
```
Comment (anchal-agrawal) [inline cmd/vicadmin/server.go]: My understanding is that if `getSessionFromRequest` returns an error here, the `if sess == nil` block below will no longer get executed and the degraded dashboard will not kick in. Just wanted to check if that's the intended behavior or not.

Also, while doing a quick test with your branch,  if I set `sess = nil` explicitly right below the `sess, err := s.getSessionFromRequest(ctx, req)` line, I get too many redirects on the browser. I'll dig deeper to see what's up.

Comment (anchal-agrawal) [inline cmd/vicadmin/server.go]: Sidenote - while I was implementing the "degraded" dashboard behavior, I ran into this exact redirect issue upon redirecting the user to the logout handler if the session was nil.

File: cmd/vicadmin/server.go (hunk 11 above)
Comment (hickeng) [inline cmd/vicadmin/server.go]: @vburenin if that's coming from vSphere, then yes. If it's from govmomi I don't think so.
@gigawhitlocks What's the real impact of redirecting to logout (and handling the case where you're not actually logged in) for ALL errors? I'm not seeing the alternative path added much value at this time.

File: cmd/vicadmin/vicadm.go (hunk 5 above)
Comment (hickeng) [inline cmd/vicadmin/vicadm.go]: @gigawhitlocks the only case where we're aiming for a degraded dashboard is if vic-admin cannot connect to vSphere to validate supplied credentials and the only thing we need to report at that time is that we cannot connect to vsphere.

The only reason to do this via dashboard was to provide some additional diagnostics as to _why_ we cannot connect but if it's causing issues then a message similar to that for internal server error suffices to let the user know why they cannot authenticate.

File: lib/vicadmin/validate.go (hunk 16 above)
Comment (hickeng) [inline lib/vicadmin/validate.go]: @gigawhitlocks Should this check go inside the block where we've confirmed that we can connect to vSphere along with the firewall checks?

**Comment by anchal-agrawal** [comment]
LGTM :+1:

**Comment by gigawhitlocks** [comment]
♻️
